### PR TITLE
streamline xds proxy logs

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/27174
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/istio-agent/xds-proxy.go [hunk 1]
```diff
@@ -166,12 +168,20 @@ func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDisc
...
+				proxyLog.Debugf("upstream terminated with status %v", err)
+			} else {
+				proxyLog.Warnf("upstream terminated with unexpected error %v", err)
+			}
 			_ = upstream.CloseSend()
 			return err
 		case err := <-downstreamError:
 			// error from downstream Envoy.
-			proxyLog.Infof("downstream terminated with status %v", err)
+			if isExpectedGRPCError(err) {
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds-proxy.go:180]: Now, i think downstream(envoy) will not reconnect periodically

Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:180]: why? I just changed the log level

File: pkg/istio-agent/xds-proxy.go [hunk 2]
```diff
@@ -166,12 +168,20 @@ func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDisc
 		select {
 		case err := <-upstreamError:
 			// error from upstream Istiod.
-			proxyLog.Infof("Upstream terminated with status %v", err)
+			if isExpectedGRPCError(err) {
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds-proxy.go:171]: For upstream error, not sure if this is right `"client disconnected"`.

Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:171]: I added "transport is closing" to `isExpectedGRPCError` method

File: pkg/istio-agent/xds-proxy.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds-proxy.go:180]: isn't the connection long running?

File: pkg/istio-agent/xds-proxy.go [hunk 3]
```diff
@@ -247,6 +257,23 @@ func (p *XdsProxy) close() {
...
+func isExpectedGRPCError(err error) bool {
+	if err == io.EOF {
+		return true
+	}
+
+	s := status.Convert(err)
+	if s.Code() == codes.Canceled || s.Code() == codes.DeadlineExceeded {
+		return true
+	}
+	if s.Code() == codes.Unavailable && (s.Message() == "client disconnected" || s.Message() == "transport is closing ") {
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds-proxy.go:271]: ` transport is closing ` has a wider meaning https://chromium.googlesource.com/external/github.com/grpc/grpc-go/#the-rpc-failed-with-error

Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:271]: Yes. It has but among all reasons , I think only #3 is applicable - not others. We will catch others much early except tcp keep alive which we do not set.  Do you see any alternative way of catching this?

**Comment by hzxuzhonghu** [comment]
I think we can log all the error, in proxy side, it is not like server. This should not be noisy.

After looking at the code, the upstream and downstream are coupled together, one side error would cause the other side to recreate a new stream.

File: pkg/istio-agent/xds-proxy.go (hunk 1 above)
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:180]: Yes it is. But I did not change those semantics now. I just condition and changed the log level. May be I am not understanding what you are saying - has this PR changed any behaviour wrt to that?

**Comment by ramaraochavali** [comment]
> After looking at the code, the upstream and downstream are coupled together, one side error would cause the other side to recreate a new stream.

Yes. That is how it is today. We need to fix that

**Review by howardjohn — APPROVED** [review]

**Pull request merged**